src/atlas/interpolation/method/sphericalvector/SphericalVector.cc

```diff
@@ -0,0 +1,296 @@
+
+  ATLAS_ASSERT(complexWeights_->nonZeros() == matrix().nonZeros());
+  ATLAS_ASSERT(realWeights_->nonZeros() == matrix().nonZeros());
+}
+
+void SphericalVector::print(std::ostream&) const { ATLAS_NOTIMPLEMENTED; }
+
+void SphericalVector::do_execute(const FieldSet& sourceFieldSet,
+                                 FieldSet& targetFieldSet,
+                                 Metadata& metadata) const {
+  ATLAS_TRACE("atlas::interpolation::method::SphericalVector::do_execute()");
+
+  const auto nFields = sourceFieldSet.size();
+  ATLAS_ASSERT(nFields == targetFieldSet.size());
+
+  for (auto i = 0; i < sourceFieldSet.size(); ++i) {
+    do_execute(sourceFieldSet[i], targetFieldSet[i], metadata);
+  }
+}
+
+void SphericalVector::do_execute(const Field& sourceField, Field& targetField,
+                                 Metadata&) const {
+  ATLAS_TRACE("atlas::interpolation::method::SphericalVector::do_execute()");
+
+  const auto fieldType = sourceField.metadata().getString("type", "");
+  if (fieldType != "vector") {
+
+    auto metadata = Metadata();
+    Method::do_execute(sourceField, targetField, metadata);
+
+    return;
+  }
+
+  if (target_.size() == 0) {
+    return;
+  }
+
+  ATLAS_ASSERT_MSG(sourceField.variables() == 2 || sourceField.variables() == 3,
+                   "Vector field can only have 2 or 3 components.");
+
+  Method::check_compatibility(sourceField, targetField, matrix());
+
+  haloExchange(sourceField);
+
+  if (sourceField.datatype().kind() == array::DataType::KIND_REAL64) {
+    interpolate_vector_field<double>(sourceField, targetField);
+  } else if (sourceField.datatype().kind() == array::DataType::KIND_REAL32) {
+    interpolate_vector_field<float>(sourceField, targetField);
+  } else {
+    ATLAS_NOTIMPLEMENTED;
+  }
+
+  targetField.set_dirty();
+}
+
+template <typename Value>
+void SphericalVector::interpolate_vector_field(const Field& sourceField,
+                                               Field& targetField) const {
+  if (sourceField.rank() == 2) {
+    interpolate_vector_field<Value, 2>(sourceField, targetField);
+  } else if (sourceField.rank() == 3) {
+    interpolate_vector_field<Value, 3>(sourceField, targetField);
+  } else {
+    ATLAS_NOTIMPLEMENTED;
+  }
+}
+
+template <typename Value, int Rank>
+void SphericalVector::interpolate_vector_field(const Field& sourceField,
+                                               Field& targetField) const {
+
+  const auto sourceView = array::make_view<Value, Rank>(sourceField);
+  auto targetView = array::make_view<Value, Rank>(targetField);
+  targetView.assign(0.);
+
+  const auto horizontalComponent = [](const auto& sourceVars, auto& targetVars,
+                                      const auto& complexWeight) {
+    const auto sourceVector = Complex(sourceVars(0), sourceVars(1));
+    const auto targetVector = complexWeight * sourceVector;
+    targetVars(0) += targetVector.real();
+    targetVars(1) += targetVector.imag();
```

Comment:
Without going into the detailed maths, why do you have to
`targetView.assign(0)`
and then use `+=` for the targetVars ? Can this not be replaced with just `=` ?

Reply:
It's because each target element (named `targetVars` here) has a contribution from multiple source elements (`sourceVars`). If I changed `+=` to `=`, it would set the target element to the last source element on that row of the matrix.

Reply:
I could do the `assign(0.)` within `matrixMultiply` function, but it would have to be on the first line, and therefore be the equivalent operation as it currently is. The `SparseMatrixForEach` is the only function that has knowledge of `i` as a whole row, but it doesn't know what `functor` is doing (it's used in a couple of different contexts).

I suspect details like this will have to be refactored to remove code duplication when the adjoint methods are added. In that context, you don't assign zero to the array you're writing to.

Reply:
OK it can stay as-is. We should not attempt premature optimisation now.